crates/biome_service/src/workspace/server.rs

```diff
@@ -458,39 +436,52 @@ impl Workspace for WorkspaceServer {
             }
         }
 
-        let index = self.set_source(source);
-        self.syntax.remove(&params.path);
-        self.documents.insert(
-            params.path.clone(),
+        if let Some(project_key) = self.path_belongs_to_other_project(&params.path) {
+            self.set_current_project(project_key);
+        }
+
+        let mut index = self.set_source(source);
+        let parsed = self.parse(&params.path, &params.content, index)?;
+
+        if let Some(language) = parsed.language {
+            index = self.set_source(language);
+        }
```

Comment:
Yeah, I agree we need to carefully think this through. In fact, with multi-file analysis we need to decide what “ignoring” a file even means. Does it mean that we should not show any diagnostics about the file, or does it mean we cannot even extract any information from it? For examples:

* A generated file may import non-generated files. If in turn a non-generated file imports the generated one, that may lead to a cycle. If we don’t analyse the generated file, we would fail to detect the cycle and cannot show the diagnostic on the non-generated file.
* `node_modules` will typically be ignored even though we’ll want to extract type information from it.

These use cases make me think we should probably parse even the ignored files.

Note this doesn’t imply we need to parse every file inside `node_modules`. I would propose we start traversing from the `included` files and expand traversal to those that get imported from them. For those we have good reason to parse them, even if we never need to get diagnostics from them.

Finally, of course there are other reasons why users may want to exclude files (I’m intentionally avoided the word “ignore” here). A file may be too big, or it might contain invalid syntax. That’s a very different reason, and indeed we would want to skip parsing altogether in such cases. I’m not sure yet what’s the best way to distinguish between these cases however…

Reply:
Oh, maybe one more point that we should consider. This code is inside the `open_file()` function. If the file should truly be excluded, I think we should not even call `open_file()` to begin with. So the intention should be, if we call `open_file()` we want to at least get some information out of it, meaning that parsing wouldn't be unnecessary.

Reply:
Great. The CLI already does that, but not the LSP. So we should update the LSP code to check if the file is ignored